### PR TITLE
RETIRE_KEYS proposal

### PR DESCRIPTION
**Pull request by DavidSchinazi**
We had discussed RETIRE_KEYS in Tokyo, and this PR attempts to write that up.

RETIRE_KEYS is an explicit signal allowing peers to agree on when to discard unused keys. It simplifies the text for discarding initial keys and allows efficiently discarding handshake keys without requiring timers or risking deadlocks (#2214).

I believe this is simpler to reason about than the KEYS_ACTIVE proposal (#2237) and safer as it makes less assumptions about the handshake protocol.

### CODE REVIEW

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -672,25 +671,24 @@ However, this does not guarantee that no further packets will need to be
...
 MAY also include PADDING frames.
 
-Once this timer expires, an endpoint MUST NOT either accept or generate new
-packets using those packet protection keys.  An endpoint can discard packet
-protection keys for that encryption level.
+Endpoints coordinate on when to discard unused keys using the RETIRE_KEYS
+frame. At a high level, each endpoint sends a RETIRE_KEYS frame when they
+are done sending using those keys; when an endpoint has both sent and received
+a RETIRE_KEYS frame for a given encryption level, it discards the corresponding
+keys. More specifically, an endpoint is done sending using a given set of keys
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: This is probably not what you intended here.  I think you mean to say that all *previous* keys are discarded.

**Comment by rpaulo** [comment]
Wouldn't it be safer to still discard initial/handshake keys after some time?  I'm worried that if an implementation blindly follows what another implementation is doing the keys might never be retired.

**Comment by kazuho** [comment]
> Wouldn't it be safer to still discard initial/handshake keys after some time?

I can see that argument for Initial keys. At the moment, we require endpoints to drop the Initial keys at the earliest moments. There's a trade-off between relying on an explicit signal (that might get missed) vs. having a uniform way of retiring keys.

OTOH, I think we should solely rely on an explicit signal for dropping handshake keys. There is no risk in keeping the key forever (because only the peer can send Handshake packets that successfully validates), and because there is a tail.

**Comment by rpaulo** [comment]
@kazuho yes, that makes sense to me.

**Comment by martinduke** [comment]
I find this proposal tp be most straightforward of the three. It is clear and unambiguous signal with a minimum-size frame that allows us to drop keys without fussing with timers or messy logic around what the peer has "proven".

**Comment by DavidSchinazi** [comment]
Closing now that we have https://github.com/quicwg/base-drafts/pull/2673

**Pull request closed without merging**